File: 212-word-search-ii/word-search-ii.cpp
class TrieNode{
public:
    unordered_map<char,TrieNode*> children;
    bool endofword=false;

    void insert(string word){
        TrieNode* root=this;
        for(auto c:word){
            if(root->children.find(c)==root->children.end()){
                root->children[c]=new TrieNode();
            }
            root=root->children[c];
        }
        root->endofword=true;
    }
};


class Solution {
public:
    int rows;
    int cols;
    vector<string> result;

    void backtrack(int i,int j,vector<vector<char>>& board, vector<string>& words,TrieNode* root,string word){
        if(i<0 || i>=rows || j<0 || j>=cols || board[i][j]=='#') return;
        char ch=board[i][j];
        if(root->children.find(ch)==root->children.end()) return;
        word.push_back(ch);
        root=root->children[ch];
        if(root->endofword){
            result.push_back(word);
            root->endofword=false;
        }
        board[i][j]='#';
        backtrack(i+1,j,board,words,root,word);
        backtrack(i-1,j,board,words,root,word);
        backtrack(i,j+1,board,words,root,word);
        backtrack(i,j-1,board,words,root,word);
        board[i][j]=ch;
        word.pop_back();

    }
    vector<string> findWords(vector<vector<char>>& board, vector<string>& words) {
        rows=board.size();
        cols=board[0].size();
        TrieNode* root=new TrieNode();
        for(auto c:words){
            root->insert(c);
        }
        for(int i=0;i<rows;i++){
            for(int j=0;j<cols;j++){
                backtrack(i,j,board,words,root,"");
            }
        }
        return result;
    }
};

File: 212-word-search-ii/word-search-ii.java
class Solution {
    class TrieNode{
        Map<Character, TrieNode> map = new HashMap<>();
        boolean flag = false;
    }
    TrieNode trieRoot;
    char[][] board; int m, n;
    List<String> output;
    boolean[][] viz;
    public List<String> findWords(char[][] board, String[] words) {
        this.board = board;
        this.m = board.length; this.n = board[0].length;
        this.trieRoot = new TrieNode();
        this.output = new ArrayList<>();

        for (String w:words) insert(w);

        for (int i=0; i<m; i++){
            for (int j=0; j<n; j++){
                viz = new boolean[m][n];
                dfs(i, j, new StringBuilder(), trieRoot);
            }
        }
        return output;
    }

    private void dfs(int i, int j, StringBuilder sb, TrieNode curr){
        if (i<0 || i>=m || j<0 || j>=n || viz[i][j]) return;
        char c = board[i][j];
        if (!curr.map.containsKey(c)) return;

        viz[i][j] = true;
        curr = curr.map.get(c);
        sb.append(c);

        if (curr.flag){
            output.add(sb.toString());
            curr.flag = false;
        }

        dfs(i+1, j, sb, curr);
        dfs(i-1, j, sb, curr);
        dfs(i, j+1, sb, curr);
        dfs(i, j-1, sb, curr);

        viz[i][j] = false;
        sb.setLength(sb.length()-1);
    }

    private void insert(String word){
        TrieNode curr = trieRoot;
        for (char c:word.toCharArray()){
            if (!curr.map.containsKey(c)) curr.map.put(c, new TrieNode());
            curr = curr.map.get(c);
        }
        curr.flag = true;
    }
}